block/submitter.go

```diff
@@ -81,96 +264,119 @@ func submitToDA[T any](
 
-	// Marshal all items once before the loop
+	// Start the retry loop
+	for retryStrategy.ShouldContinue() {
+		if err := waitForBackoffOrContext(ctx, retryStrategy.backoff); err != nil {
+			return err
+		}
+
+		res := attemptSubmission(ctx, m, marshaled, retryStrategy.gasPrice)
+		outcome := handleSubmissionResult(m, ctx, res, remaining, marshaled, retryStrategy, postSubmit, itemType)
+
+		remaining = outcome.RemainingItems
+		marshaled = outcome.RemainingMarshal
+		numSubmitted += outcome.NumSubmitted
+
+		if outcome.AllSubmitted {
+			return nil
+		}
+
+		retryStrategy.NextAttempt()
+	}
+
+	return createFailureError(itemType, numSubmitted, len(remaining), retryStrategy.attempt)
+}
+
+func marshalItems[T any](
+	items []T,
+	marshalFn func(T) ([]byte, error),
+	itemType string,
+) ([][]byte, error) {
 	marshaled := make([][]byte, len(items))
 	for i, item := range items {
 		bz, err := marshalFn(item)
 		if err != nil {
-			return fmt.Errorf("failed to marshal item: %w", err)
+			return nil, fmt.Errorf("failed to marshal %s item: %w", itemType, err)
 		}
 		marshaled[i] = bz
 	}
-	remLen := len(items)
 
-	for !submittedAll && attempt < maxSubmitAttempts {
-		select {
-		case <-ctx.Done():
-			m.logger.Info("context done, stopping submission loop")
-			return nil
-		case <-time.After(backoff):
-		}
+	return marshaled, nil
+}
 
-		// Use the current remaining items and marshaled bytes
-		currMarshaled := marshaled
-		remLen = len(remaining)
+func waitForBackoffOrContext(ctx context.Context, backoff time.Duration) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-time.After(backoff):
+		return nil
+	}
+}
 
-		submitctx, submitCtxCancel := context.WithTimeout(ctx, 60*time.Second)
+func attemptSubmission(
+	ctx context.Context,
+	m *Manager,
+	marshaled [][]byte,
+	gasPrice float64,
+) coreda.ResultSubmit {
+	submitCtx, cancel := context.WithTimeout(ctx, submissionTimeout)
+	defer cancel()
 
-		// Record DA submission retry attempt
-		m.recordDAMetrics("submission", DAModeRetry)
+	m.recordDAMetrics("submission", DAModeRetry)
+	return types.SubmitWithHelpers(submitCtx, m.da, m.logger, marshaled, gasPrice, nil)
+}
 
-		res := types.SubmitWithHelpers(submitctx, m.da, m.logger, currMarshaled, gasPrice, nil)
-		submitCtxCancel()
+func handleSubmissionResult[T any](
+	m *Manager,
+	ctx context.Context,
+	res coreda.ResultSubmit,
+	remaining []T,
+	marshaled [][]byte,
+	retryStrategy *RetryStrategy,
+	postSubmit func([]T, *coreda.ResultSubmit, float64),
+	itemType string,
+) SubmissionOutcome[T] {
+	switch res.Code {
+	case coreda.StatusSuccess:
+		return handleSuccessfulSubmission(m, remaining, marshaled, &res, postSubmit, retryStrategy, itemType)
 
-		switch res.Code {
-		case coreda.StatusSuccess:
-			// Record successful DA submission
-			m.recordDAMetrics("submission", DAModeSuccess)
+	case coreda.StatusNotIncludedInBlock, coreda.StatusAlreadyInMempool:
+		handleMempoolFailure(m, &res, retryStrategy, retryStrategy.attempt)
```

Comment:
nit: to be consistent, it's likely cleaner to just make SubmissionOutcome a return type of handleMempoolFailure and return that, just like it's done for handleSuccessfulSubmission and handleTooBigError.